cairo/ethereum/cancun/trie.cairo

```diff
@@ -372,55 +464,353 @@ func bytes_to_nibble_list{bitwise_ptr: BitwiseBuiltin*}(bytes_: Bytes) -> Bytes
-//     // return ExtensionNode(prefix, encode_internal_node(patricialize(obj, level + prefix_length)))
-//         // prefix = arbitrary_key[level:level + prefix_length]
-//         // return ExtensionNode(prefix, encode_internal_node(patricialize(obj, level + prefix_length)))
-//     // branches: List[MutableMapping[Bytes, Bytes]] = []
-//     // for _ in range(16):
-//     // branches.append({})
-//         // branches.append({})
-//     // value = b''
-//     // for key in obj:
-//     // if len(key) == level:
-//     // if isinstance(obj[key], (Account, Receipt, Uint)):
-//     // raise AssertionError
-//     // value = obj[key]
-//     // else:
-//     // branches[key[level]][key] = obj[key]
-//         // if len(key) == level:
-//         // if isinstance(obj[key], (Account, Receipt, Uint)):
-//         // raise AssertionError
-//         // value = obj[key]
-//         // else:
-//         // branches[key[level]][key] = obj[key]
-//             // if isinstance(obj[key], (Account, Receipt, Uint)):
-//             // raise AssertionError
-//                 // raise AssertionError
-//             // value = obj[key]
-//         // else:
-//             // branches[key[level]][key] = obj[key]
-//     // return BranchNode([encode_internal_node(patricialize(branches[k], level + 1)) for k in range(16)], value)
-// }
+func _search_common_prefix_length{
+    range_check_ptr, substring: Bytes, level: Uint, dict_ptr_stop: BytesBytesDictAccess*
+}(obj: BytesBytesDictAccess*, current_length: felt) -> felt {
+    alloc_locals;
+    if (obj == dict_ptr_stop) {
+        return current_length;
+    }
+
+    tempvar sliced_key = Bytes(
+        new BytesStruct(obj.key.value.data + level.value, obj.key.value.len - level.value)
+    );
+    let result = common_prefix_length(substring, sliced_key);
+    let current_length = min(result, current_length);
+    if (current_length == 0) {
+        return 0;
+    }
+
+    return _search_common_prefix_length(obj + BytesBytesDictAccess.SIZE, current_length);
+}
+
+// @dev Iterate over the DictAccesses from dict_ptr_start to dict_ptr to create a new DictAccess*
+// @dev whose key[level] = nibble.
+func _get_branche_for_nibble_at_level(obj: MappingBytesBytes, nibble: felt, level: felt) -> (
+    MappingBytesBytes, Bytes
+) {
+    alloc_locals;
+    let (local branch: BytesBytesDictAccess*) = alloc();
+    local dict_ptr_stop: BytesBytesDictAccess* = obj.value.dict_ptr;
+    local value: Bytes;
+    local value_set: felt;
+
+    tempvar branch = branch;
+    tempvar dict_ptr = obj.value.dict_ptr_start;
+
+    loop:
+    let branch = cast([ap - 2], BytesBytesDictAccess*);
+    let dict_ptr = cast([ap - 1], BytesBytesDictAccess*);
+    // The verifier just needs to make sure that whatever case we are in is properly asserted.
+    tempvar is_nibble_case = nondet %{ memory.get(ids.dict_ptr.key.value.data + ids.level) == ids.nibble %};
+    tempvar is_value_case = nondet %{ int(ids.dict_ptr.key.value.len == ids.level) %};
```

Comment:
but what does this change? because each branch checks its own condition, so if the prover wants to go in `value_case` while it's not possible, it won't be able to satisfy the constraint `assert dict_ptr.key.value.len = level;`

Reply:
note that `is_zero(a-b)` would have done it as well. Overall I could remove these two hints, handle the value_case first, and otherwise do the check nibble or not

Reply:
I think it's best removing the hints and just do a is_zero
When it comes to values provided through hints, I believe the habit of having explicit and clear assertions outweighs  the risk of "over-optimizing" by relying on other implied assumptions.

In that case I prefer seeing:
```
res_0 = %{something happening here%}
res_1 = %{..%}
assert res_0 * res_1  = 1
```
which clearly outlines the valid states: either both are 0, or one of them is 0, but having both of them == 1 is an invalid state